integration-tests/tests/src/tests/sync/mongo-db-client.ts

```diff
@@ -0,0 +1,188 @@
+      // Used as a flag for knowing when to stop watching.
+      const lastDocument = { _id: numInserts, isLast: true };
+
+      function assertIsInsert<
+        T extends Document = TestDocument,
+      >(event: ChangeEvent<T>): asserts event is InsertEvent<T> {
+        if (event.operationType !== "insert") {
+          throw new Error(`Expected an insert event, got ${event}.`);
+        }
+      }
+
+      async function insertDocumentsOneByOne(): Promise<void> {
+        // There is a race with creating the watch() streams, since they won't
+        // see inserts from before they are created.
+        // Wait 500ms (490+10) before first insert to try to avoid it.
+        await sleep(490);
+        for (let i = 0; i < numInserts; i++) {
+          await collection.insertOne({ _id: i, text: text });
+          await sleep(10);
+        }
+        await collection.insertOne(lastDocument);
+      }
+
+      it("streams any inserted documents", async function (this: AppContext & Mocha.Context) {
+        await Promise.all([
+          insertDocumentsOneByOne(),
+          (async () => {
+            let expectedId = 0;
+            for await (const event of collection.watch()) {
+              assertIsInsert(event);
+              expect(event.fullDocument._id).to.equal(expectedId++);
+              if (event.fullDocument.isLast) break;
+            }
+          })(),
+        ]);
+      });
+
+      it("streams inserted documents using 'filter' option", async function (this: AppContext & Mocha.Context) {
+        await Promise.all([
+          insertDocumentsOneByOne(),
+          (async () => {
+            const watchedId = 3;
+            const filter = {
+              $or: [{ "fullDocument._id": watchedId, "fullDocument.text": text }, { "fullDocument.isLast": true }],
+            };
+            let seenIt = false;
+            for await (const event of collection.watch({ filter })) {
+              assertIsInsert(event);
+              if (event.fullDocument.isLast) break;
+              expect(event.fullDocument._id).to.equal(watchedId);
+              seenIt = true;
+            }
+            expect(seenIt).to.be.true;
+          })(),
+        ]);
+      });
+
+      it("streams inserted documents using 'ids' option", async function (this: AppContext & Mocha.Context) {
+        await Promise.all([
+          insertDocumentsOneByOne(),
+          (async () => {
+            const watchedId = 3;
+            let seenIt = false;
+            for await (const event of collection.watch({ ids: [watchedId, lastDocument._id] })) {
+              assertIsInsert(event);
+              if (event.fullDocument.isLast) break;
+              expect(event.fullDocument._id).to.equal(watchedId);
+              seenIt = true;
+            }
+            expect(seenIt).to.be.true;
+          })(),
+        ]);
+      });
+
+      it("throws when the user is logged out", async function (this: AppContext & Mocha.Context) {
```

Comment:
This test needs some TLC.

* `this.app.currentUser` seem to be `null` here.
* We're not hitting the `expect.fail()` (which we shouldn't), but we're also not entering the `catch` block (which we should).
* Need to consider how to handle potential tests added after this one if this test logs the user out.
* We do get a `401` auth error if e.g. trying to log in in the `finally` clause.

Reply:
We have the same challenge with Realm Web: https://github.com/realm/realm-js/blob/master/packages/realm-web-integration-tests/src/remote-mongodb-service.test.ts#L417

Hopefully @kraenhansen remembers why